File: views/src/main/java/ke/co/toshngure/views/media/NetworkImage.java
/*
 * Copyright (c) 2018.
 *
 * Anthony Ngure
 *
 * Email : anthonyngure25@gmail.com
 */

package ke.co.toshngure.views.media;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.util.AttributeSet;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.ProgressBar;

import androidx.annotation.ColorInt;
import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.content.res.AppCompatResources;
import androidx.appcompat.widget.AppCompatImageView;

import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;

import java.io.File;
import java.lang.ref.WeakReference;

import ke.co.toshngure.basecode.app.GlideRequests;
import ke.co.toshngure.views.CircleImageView;
import ke.co.toshngure.views.R;


/**
 * Created by Anthony Ngure on 20/02/2017.
 * Email : anthonyngure25@gmail.com.
 */

public class NetworkImage extends FrameLayout {

    private static final String TAG = NetworkImage.class.getSimpleName();

    protected ImageView mImageView;
    protected ImageView mBackgroundImageView;
    protected ImageView mErrorButton;
    protected ProgressBar mProgressBar;
    private LoadingCallBack loadingCallBack;
    private boolean noRetry = false;

    @Nullable
    private Drawable placeholder = new ColorDrawable(Color.LTGRAY);

    @Nullable
    private Drawable error = new ColorDrawable(Color.LTGRAY);

    public NetworkImage(Context context) {
        this(context, null);
    }

    public NetworkImage(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public NetworkImage(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        LayoutInflater.from(context).inflate(R.layout.view_network_image, this, true);

        mErrorButton = findViewById(R.id.errorButton);
        mProgressBar = findViewById(R.id.progressBar);

        FrameLayout imageFL = findViewById(R.id.imageFL);
        TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.NetworkImage);

        boolean circled = typedArray.getBoolean(R.styleable.NetworkImage_niCircled, false);
        FrameLayout.LayoutParams layoutParams = new LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT);
        if (circled) {
            CircleImageView imageView = new CircleImageView(context);
            int borderColor = typedArray.getColor(R.styleable.NetworkImage_niBorderColor, Color.WHITE);
            int borderWidth = typedArray.getDimensionPixelOffset(R.styleable.NetworkImage_niBorderWidth, 0);
            imageView.setBorderColor(borderColor);
            imageView.setBorderWidth(borderWidth);
            imageView.setBorderOverlay(true);
            mImageView = imageView;
            // mImageView = new CircleImageView(context);
            mBackgroundImageView = new CircleImageView(context);
        } else {
            mImageView = new AppCompatImageView(context);
            mBackgroundImageView = new AppCompatImageView(context);
            mImageView.setScaleType(ImageView.ScaleType.CENTER_CROP);
            mBackgroundImageView.setScaleType(ImageView.ScaleType.CENTER_CROP);
        }
        imageFL.addView(mBackgroundImageView, layoutParams);
        imageFL.addView(mImageView, layoutParams);

        /*Image*/
        int drawableResId = typedArray.getResourceId(R.styleable.NetworkImage_niSrc, -1);
        if (drawableResId != -1) {
            Drawable drawable = AppCompatResources.getDrawable(getContext(), drawableResId);
            setImageDrawable(drawable);
        }

        /*Background*/
        int backgroundResId = typedArray.getResourceId(R.styleable.NetworkImage_niBackground, -1);
        if (backgroundResId != -1) {
            Drawable backgroundDrawable = AppCompatResources.getDrawable(getContext(), backgroundResId);
            mBackgroundImageView.setImageDrawable(backgroundDrawable);
        }
        typedArray.recycle();

    }

    public void loadImageFromNetwork(final String networkPath, GlideRequests glideRequests) {
        mProgressBar.setVisibility(VISIBLE);
        mErrorButton.setOnClickListener(v -> {
            Log.d(TAG, "Retrying to loadFromNetwork image");
            mProgressBar.setVisibility(VISIBLE);
            mErrorButton.setVisibility(GONE);
            loadImageFromNetwork(networkPath, glideRequests);
        });

        glideRequests.load(networkPath)
                .centerCrop()
                .placeholder(placeholder)
                .error(error)
                .listener(new Listener(mImageView, mProgressBar, mErrorButton, loadingCallBack, noRetry))
                .into(mImageView);
    }

    public void loadImageFromMediaStore(String path, GlideRequests glideRequests) {
        mProgressBar.setVisibility(VISIBLE);
        File file = new File(path);
        glideRequests.load(Uri.fromFile(file))
                .centerCrop()
                .placeholder(new ColorDrawable(Color.LTGRAY))
                .error(R.drawable.ic_place_holder)
                .listener(new Listener(mImageView, mProgressBar, mErrorButton, loadingCallBack, noRetry))
                .into(mImageView);
    }


    public void loadImageFromMediaStore(Uri uri, GlideRequests glideRequests) {
        mProgressBar.setVisibility(VISIBLE);
        glideRequests.load(uri)
                .centerCrop()
                .placeholder(new ColorDrawable(Color.LTGRAY))
                .error(R.drawable.ic_place_holder)
                .listener(new Listener(mImageView, mProgressBar, mErrorButton, loadingCallBack, noRetry))
                .into(mImageView);
    }

    public NetworkImage setLoadingCallBack(LoadingCallBack loadingCallBack) {
        this.loadingCallBack = loadingCallBack;
        return this;
    }

    public NetworkImage setImageDrawable(Drawable drawable) {
        mImageView.setImageDrawable(drawable);
        mImageView.setVisibility(VISIBLE);
        mProgressBar.setVisibility(GONE);
        mErrorButton.setVisibility(GONE);
        return this;
    }

    public NetworkImage setImageUri(Uri uri) {
        mImageView.setImageURI(uri);
        mImageView.setVisibility(VISIBLE);
        mProgressBar.setVisibility(GONE);
        mErrorButton.setVisibility(GONE);
        return this;
    }

    public ImageView getImageView() {
        return this.mImageView;
    }


    public NetworkImage setImageBackgroundColor(@ColorInt int color) {
        mBackgroundImageView.setImageDrawable(new ColorDrawable(color));
        return this;
    }


    public NetworkImage setImageResource(@DrawableRes int resId) {
        Drawable drawable = AppCompatResources.getDrawable(getContext(), resId);
        mImageView.setImageDrawable(drawable);
        return this;
    }

    public NetworkImage setPlaceholderDrawable(@Nullable Drawable drawable) {
        this.placeholder = drawable;
        // 0702 238 023
        return this;
    }

    public NetworkImage setErrorDrawable(@Nullable Drawable drawable) {
        this.error = drawable;
        return this;
    }

    public NetworkImage setNoRetry() {
        this.noRetry = true;
        return this;
    }


    public NetworkImage setScaleType(ImageView.ScaleType scaleType) {
        mImageView.setScaleType(scaleType);
        return this;
    }


    public interface LoadingCallBack {
        void onSuccess(Drawable drawable);
    }

    /**
     * Glide Callback which clears the ImageView's background onSuccess. This is done to reduce
     * overdraw. A weak reference is used to avoid leaking the Activity context because the Callback
     * will be strongly referenced by Glide.
     */
    static class Listener implements RequestListener<Drawable> {

        final WeakReference<ImageView> imageViewWeakReference;
        final WeakReference<ProgressBar> progressBarWeakReference;
        final WeakReference<ImageView> errorImageViewWeakReference;
        final WeakReference<Boolean> noRetryWeakReference;
        final WeakReference<LoadingCallBack> loadingCallBackWeakReference;

        private Listener(ImageView imageView, ProgressBar progressBar, ImageView errorImageView,
                         LoadingCallBack loadingCallBack, Boolean noRetry) {
            imageViewWeakReference = new WeakReference<>(imageView);
            progressBarWeakReference = new WeakReference<>(progressBar);
            errorImageViewWeakReference = new WeakReference<>(errorImageView);
            loadingCallBackWeakReference = new WeakReference<>(loadingCallBack);
            noRetryWeakReference = new WeakReference<>(noRetry);
        }


        @Override
        public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
            Log.e(TAG, String.valueOf(e));

            ProgressBar progressBar = progressBarWeakReference.get();
            if (progressBar != null) {
                progressBar.setVisibility(GONE);
            }

            ImageView errorImageView = errorImageViewWeakReference.get();

            boolean noRetry = noRetryWeakReference.get();
            if (errorImageView != null) {
                errorImageView.setVisibility(noRetry ? GONE : VISIBLE);
            }

            return false;
        }

        @Override
        public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
            final ImageView imageView = imageViewWeakReference.get();
            LoadingCallBack loadingCallBack = loadingCallBackWeakReference.get();
            if (loadingCallBack != null) {
                loadingCallBack.onSuccess(resource);
            }
            ProgressBar progressBar = progressBarWeakReference.get();
            if (progressBar != null) {
                progressBar.setVisibility(GONE);
            }

            ImageView errorImageView = errorImageViewWeakReference.get();
            if (errorImageView != null) {
                errorImageView.setVisibility(GONE);
            }
            return false;
        }


    }

}

File: views/src/main/java/ke/co/toshngure/views/media/NetworkImage.kt
/*
 * Copyright (c) 2018.
 *
 * Anthony Ngure
 *
 * Email : anthonyngure25@gmail.com
 */
package ke.co.toshngure.views.media

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.net.Uri
import android.util.AttributeSet
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.ImageView.ScaleType
import android.widget.ProgressBar
import androidx.annotation.ColorInt
import androidx.annotation.DrawableRes
import androidx.appcompat.content.res.AppCompatResources
import androidx.appcompat.widget.AppCompatImageView
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.target.Target
import ke.co.toshngure.basecode.app.GlideRequests
import ke.co.toshngure.views.CircleImageView
import ke.co.toshngure.views.R
import java.io.File
import java.lang.ref.WeakReference

/**
 * Created by Anthony Ngure on 20/02/2017.
 * Email : anthonyngure25@gmail.com.
 */
class NetworkImage @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : FrameLayout(context, attrs, defStyleAttr) {
    var imageView: ImageView
        protected set
    protected var mBackgroundImageView: ImageView
    protected var mErrorButton: ImageView
    protected var mProgressBar: ProgressBar
    private var loadingCallBack: LoadingCallBack? = null
    private var noRetry = false
    private var placeholder: Drawable? = ColorDrawable(Color.LTGRAY)
    private var error: Drawable? = ColorDrawable(Color.LTGRAY)
    fun loadImageFromNetwork(networkPath: String?, glideRequests: GlideRequests) {
        mProgressBar.visibility = VISIBLE
        mErrorButton.setOnClickListener { v: View? ->
            Log.d(TAG, "Retrying to loadFromNetwork image")
            mProgressBar.visibility = VISIBLE
            mErrorButton.visibility = GONE
            loadImageFromNetwork(networkPath, glideRequests)
        }

        glideRequests.load(networkPath)
            .centerCrop()
            .placeholder(placeholder)
            .error(error)
            .listener(Listener(imageView, mProgressBar, mErrorButton, loadingCallBack, noRetry))
            .into(imageView)
    }

    fun loadImageFromMediaStore(path: String?, glideRequests: GlideRequests) {
        mProgressBar.visibility = VISIBLE

        path?.let {
            val file = File(path)
            glideRequests.load(Uri.fromFile(file))
                .centerCrop()
                .placeholder(ColorDrawable(Color.LTGRAY))
                .error(R.drawable.ic_place_holder)
                .listener(Listener(imageView, mProgressBar, mErrorButton, loadingCallBack, noRetry))
                .into(imageView)
        }
    }

    fun loadImageFromMediaStore(uri: Uri?, glideRequests: GlideRequests) {
        mProgressBar.visibility = VISIBLE
        glideRequests.load(uri)
            .centerCrop()
            .placeholder(ColorDrawable(Color.LTGRAY))
            .error(R.drawable.ic_place_holder)
            .listener(Listener(imageView, mProgressBar, mErrorButton, loadingCallBack, noRetry))
            .into(imageView)
    }

    fun setLoadingCallBack(loadingCallBack: LoadingCallBack?): NetworkImage {
        this.loadingCallBack = loadingCallBack
        return this
    }

    fun setImageDrawable(drawable: Drawable?): NetworkImage {
        imageView.setImageDrawable(drawable)
        imageView.visibility = VISIBLE
        mProgressBar.visibility = GONE
        mErrorButton.visibility = GONE
        return this
    }

    fun setImageUri(uri: Uri?): NetworkImage {
        imageView.setImageURI(uri)
        imageView.visibility = VISIBLE
        mProgressBar.visibility = GONE
        mErrorButton.visibility = GONE
        return this
    }

    fun setImageBackgroundColor(@ColorInt color: Int): NetworkImage {
        mBackgroundImageView.setImageDrawable(ColorDrawable(color))
        return this
    }

    fun setImageResource(@DrawableRes resId: Int): NetworkImage {
        val drawable = AppCompatResources.getDrawable(context, resId)
        imageView.setImageDrawable(drawable)
        return this
    }

    fun setPlaceholderDrawable(drawable: Drawable?): NetworkImage {
        placeholder = drawable
        // 0702 238 023
        return this
    }

    fun setErrorDrawable(drawable: Drawable?): NetworkImage {
        error = drawable
        return this
    }

    fun setNoRetry(): NetworkImage {
        noRetry = true
        return this
    }

    fun setScaleType(scaleType: ScaleType?): NetworkImage {
        imageView.scaleType = scaleType
        return this
    }

    interface LoadingCallBack {
        fun onSuccess(drawable: Drawable?)
    }

    /**
     * Glide Callback which clears the ImageView's background onSuccess. This is done to reduce
     * overdraw. A weak reference is used to avoid leaking the Activity context because the Callback
     * will be strongly referenced by Glide.
     */
    internal class Listener(
        imageView: ImageView?, progressBar: ProgressBar, errorImageView: ImageView,
        loadingCallBack: LoadingCallBack?, noRetry: Boolean
    ) : RequestListener<Drawable?> {
        val imageViewWeakReference: WeakReference<ImageView?>
        val progressBarWeakReference: WeakReference<ProgressBar>
        val errorImageViewWeakReference: WeakReference<ImageView>
        val noRetryWeakReference: WeakReference<Boolean>
        val loadingCallBackWeakReference: WeakReference<LoadingCallBack?>
        override fun onLoadFailed(
            e: GlideException?,
            model: Any,
            target: Target<Drawable?>,
            isFirstResource: Boolean
        ): Boolean {
            Log.e(TAG, "error")
            Log.e(TAG, e.toString())
            val progressBar = progressBarWeakReference.get()
            if (progressBar != null) {
                progressBar.visibility = GONE
            }
            val errorImageView = errorImageViewWeakReference.get()
            val noRetry = noRetryWeakReference.get()!!
            if (errorImageView != null) {
                errorImageView.visibility = if (noRetry) GONE else VISIBLE
            }
            return false
        }

        override fun onResourceReady(
            resource: Drawable?,
            model: Any,
            target: Target<Drawable?>,
            dataSource: DataSource,
            isFirstResource: Boolean
        ): Boolean {
            val imageView = imageViewWeakReference.get()
            val loadingCallBack = loadingCallBackWeakReference.get()
            loadingCallBack?.onSuccess(resource)
            val progressBar = progressBarWeakReference.get()
            if (progressBar != null) {
                progressBar.visibility = GONE
            }
            val errorImageView = errorImageViewWeakReference.get()
            if (errorImageView != null) {
                errorImageView.visibility = GONE
            }
            return false
        }

        init {
            imageViewWeakReference = WeakReference(imageView)
            progressBarWeakReference = WeakReference(progressBar)
            errorImageViewWeakReference = WeakReference(errorImageView)
            loadingCallBackWeakReference = WeakReference(loadingCallBack)
            noRetryWeakReference = WeakReference(noRetry)
        }
    }

    companion object {
        private val TAG = NetworkImage::class.java.simpleName
    }

    init {
        LayoutInflater.from(context).inflate(R.layout.view_network_image, this, true)
        mErrorButton = findViewById(R.id.errorButton)
        mProgressBar = findViewById(R.id.progressBar)
        val imageFL = findViewById<FrameLayout>(R.id.imageFL)
        val typedArray = context.obtainStyledAttributes(attrs, R.styleable.NetworkImage)
        val circled = typedArray.getBoolean(R.styleable.NetworkImage_niCircled, false)
        val layoutParams = LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT
        )
        if (circled) {
            val imageView = CircleImageView(context)
            val borderColor =
                typedArray.getColor(R.styleable.NetworkImage_niBorderColor, Color.WHITE)
            val borderWidth =
                typedArray.getDimensionPixelOffset(R.styleable.NetworkImage_niBorderWidth, 0)
            imageView.borderColor = borderColor
            imageView.borderWidth = borderWidth
            imageView.isBorderOverlay = true
            this.imageView = imageView
            // mImageView = new CircleImageView(context);
            mBackgroundImageView = CircleImageView(context)
        } else {
            imageView = AppCompatImageView(context)
            mBackgroundImageView = AppCompatImageView(context)
            imageView.setScaleType(ScaleType.CENTER_CROP)
            mBackgroundImageView.setScaleType(ScaleType.CENTER_CROP)
        }
        imageFL.addView(mBackgroundImageView, layoutParams)
        imageFL.addView(imageView, layoutParams)

        /*Image*/
        val drawableResId = typedArray.getResourceId(R.styleable.NetworkImage_niSrc, -1)
        if (drawableResId != -1) {
            val drawable = AppCompatResources.getDrawable(getContext(), drawableResId)
            setImageDrawable(drawable)
        }

        /*Background*/
        val backgroundResId = typedArray.getResourceId(R.styleable.NetworkImage_niBackground, -1)
        if (backgroundResId != -1) {
            val backgroundDrawable = AppCompatResources.getDrawable(getContext(), backgroundResId)
            mBackgroundImageView.setImageDrawable(backgroundDrawable)
        }
        typedArray.recycle()
    }
}